Refactor code

rowan08 opened the pull request:
Refactored code based on agreed guidelines. Added guidelines to README file.
Guidelines followed as per https://github.com/davidbrownza/CORMAP/issues/14
Also changed `int `to `unsigned int` in some cases to remove compilation warnings.
Code tested by running make.
Issue: https://github.com/davidbrownza/CORMAP/issues/34

rowan08 committed:
Refactored code based on agreed guidelines. Added guidelines to README file.
Guidelines followed as per https://github.com/davidbrownza/CORMAP/issues/14
Also changed int to unsigned int in some cases to remove compilation warnings.
Code tested by runnning make.
Issue: https://github.com/davidbrownza/CORMAP/issues/34

rowan08 committed:
Update README.md

rowan08 committed:
Update README.md

davidbrownza commented on README.md:
```diff
@@ -5,3 +5,22 @@ Simple C++ classes for Object Relational Mapping
+1. All private member variables get an underscore in front of them.
+   * Users cannot access these variables directly, so rather they be ugly than the method parameters.
+   * Private member functions are exceptions to this.
+2. Avoid the use of `this->` and `this`.
```
Maybe just say `this`

rowan08 replied:
Done

davidbrownza commented on src/conn/DBConnection.cpp:
```diff
@@ -12,41 +12,33 @@
-    prep_stmt = conn->prepareStatement(sql);
-    
+int DBConnection::prepareStatement(string sql) {    
+    _prep_stmt = _conn->prepareStatement(sql);
```
Just thought of another one. The variable names should use camel case i.e. this should be `_prepStmt`. The way it currently is is more of a Python convention.

rowan08 replied:
Apparently, it's called snake_case (which seems appropriate)

rowan08 replied:
Done

davidbrownza commented on README.md:
```diff
@@ -5,3 +5,22 @@ Simple C++ classes for Object Relational Mapping
+
+### Developer guidelines
+Cormap has been developed using the following code guidelines:
+
```
First rule should be: "Use spaces, not tabs."

davidbrownza replied:
:P

rowan08 replied:
Done

davidbrownza replied:
Also add the boolean naming convention to this list

rowan08 replied:
Done

davidbrownza commented on src/conn/DBConnection.h:
```diff
@@ -30,10 +30,10 @@ class DBConnection
-        sql::Connection *conn;
-        sql::PreparedStatement *prep_stmt;
+        sql::Connection *_conn;
+        sql::PreparedStatement *_prep_stmt;
```
Camel case: `_prepStmt`

rowan08 replied:
Can we assume that _prepStmt and _conn are suitable shorthand?
Or do these need to be changed to _preparedStatement and _connection? 
(because "prettiness")

rowan08 replied:
Done as `_prepStmt` for now

davidbrownza replied:
Hmmm in my comment for the protected variable in the Model class, I said use '_connection', but I think it is fine to leave these here.

davidbrownza replied:
I don't know, for consistency, maybe they should change...

Up to you.

rowan08 replied:
I agree. I am happy to change them.
One question though. Will the `_connection` variable for this class clash with the `_connection` variable from the model class, or are they protected by scope?

davidbrownza replied:
They are protected by scope. Even if they were both public, they wouldn't clash as they are class attributes, not global variables i.e. DBConnection._connection cannot clash with Model._connection

rowan08 replied:
Done

davidbrownza commented on src/fields/BooleanField.cpp:
```diff
@@ -1,7 +1,7 @@
 
-BooleanField::BooleanField(string name, bool defaultValue, bool primary, bool unique, bool nullable): 
-    Field(name, BOOLEAN, primary, unique, nullable) { 
+BooleanField::BooleanField(string fieldName, bool defaultValue, bool primaryKey, bool unique, bool nullable): 
```
Hmmm now that I think about it, booleans are normally named like so:
`isPrimaryKey`
`isUnique`
`isNullable`
Not sure if it it is worth the effort though...

davidbrownza replied:
Up to you

rowan08 replied:
It is a nice naming convention. Plus it's still fairly early days, so I don't mind putting in a bit of extra work for stuff like this (while it's relatively easy to do so without breaking all the things).

rowan08 replied:
Done

davidbrownza commented on src/fields/BooleanField.h:
```diff
@@ -11,7 +11,7 @@ class BooleanField: public Field {
         
     public:
-        BooleanField(string name, bool defaultValue=false, bool primary=false, bool unique=false, bool nullable=true);
+        BooleanField(string fieldName, bool defaultValue=false, bool primaryKey=false, bool unique=false, bool nullable=true);
```
If you do decide to change the boolean variable names, don't do it with `defaultValue` here.

rowan08 replied:
Noted

davidbrownza commented on src/fields/IntegerField.cpp:
```diff
@@ -1,16 +1,16 @@
+IntegerField::IntegerField(string fieldName, int defaultValue, bool primaryKey, bool unique, bool nullable, bool autoIncrement): 
+    Field(fieldName, INTEGER, primaryKey, unique, nullable, autoIncrement) { 
+    _fieldValue = defaultValue;
+    _autoIncrement = autoIncrement; //Should this be passed to Field (above)?
```
Yes (I think). But take out that comment now :P

rowan08 replied:
Just checking. My thinking is that:
1) We're creating a new variable _autoIncrement for the IntegerField class when 
2) The Field class already has a _autoFilled variable (which autoIncrement is passed to).
... which seemed a bit weird. Will remove the comment, though.

rowan08 replied:
Done

davidbrownza replied:
Yeah, the reason I say this is that for the integer field, autoincrement means that the field will be automatically filled. So if autoincrement is set, autofilled should be as well.

davidbrownza commented on src/model/Model.cpp:
```diff
@@ -6,71 +6,63 @@ Model::Model(string name) {
+    unsigned int size = models.size();
     
-    if(size < batchsize)
+    if (size < batchsize)
```
Put curly brackets around this if-statement and add that brackets should always be used (even when not necessary) to the README.

rowan08 replied:
Done

davidbrownza commented on src/model/Model.cpp:
```diff
@@ -6,71 +6,63 @@ Model::Model(string name) {
+    unsigned int finalFullBatch = size - rem;    
     
     //generate SQL 
     string sqlHead; // INSERT INTO table_name (fields) VALUES
```
Take out this comment

rowan08 replied:
Done (I think I removed the correct comment)
Also removed similar types of comments on the two lines following this one

davidbrownza commented on src/model/Model.cpp:
```diff
@@ -179,8 +163,7 @@ int Model::truncate() {
-int Model::setConnection(DBConnection connection)
-{    
+int Model::setConnection(DBConnection connection) {    
     conn = connection;
```
This should be `_connection`. Protected variables should be treated as private.

davidbrownza replied:
Remember to change this one.

rowan08 replied:
Done

davidbrownza commented on src/model/Model.cpp:
```diff
@@ -198,15 +181,13 @@ void Model::generateSQLParts(string &sqlHead, string &sqlBatch, string &sqlUpdat
+            if (firstField) {
                 firstField = false;
                 
                 sqlFields +=  f->getName();
```
I may be wrong here, but it looks like there is an extra space after the `+=` here.

rowan08 replied:
Well spotted. Fixed.
Also found some additional places this happened.

rowan08 committed:
Changes after CR:
Added additional developer guidelines to README file.
Renamed additional variable according to agreed conventions.
Fixed some spacing issues and removed unwanted comments.
Tested with make
Issue: https://github.com/davidbrownza/CORMAP/issues/34

rowan08 committed:
Update README.md

rowan08 commented on src/fields/Field.cpp:
```diff
@@ -1,36 +1,36 @@
-    uniqueValue = unique;
-    nullable = null;
-    autoFilled = autoF;
+Field::Field(string fieldName, FieldType fieldType, bool isPrimaryKey, bool isUnique, bool isNullable, bool autoFilled) {
```
**bool check:** `isAutoFilled`?

davidbrownza replied:
Yup

davidbrownza replied:
Yup

rowan08 replied:
Done

rowan08 commented on src/fields/IntegerField.cpp:
```diff
@@ -1,16 +1,16 @@
-    fieldValue = defaultValue;
-    autoIncrement = autoInc;
+IntegerField::IntegerField(string fieldName, int defaultValue, bool isPrimaryKey, bool isUnique, bool isNullable, bool autoIncrement): 
+    Field(fieldName, INTEGER, isPrimaryKey, isUnique, isNullable, autoIncrement) { 
```
**bool check:** `isAutoIncremented`?

davidbrownza replied:
Yup

davidbrownza replied:
Yup

rowan08 replied:
Done

rowan08 commented on src/model/Model.cpp:
```diff
@@ -6,93 +6,85 @@ Model::Model(string name) {
+    _connection.prepareStatement(sqlHead + sqlTail);
     
     int paramNum = 0;
     bool finalBatch = false;
```
**bool check:** `isFinalBatch`?

davidbrownza replied:
Yup

rowan08 replied:
Done

rowan08 commented on src/model/Model.cpp:
```diff
@@ -6,93 +6,85 @@ Model::Model(string name) {
+        for (unsigned int j = 0; j < m->fields.size(); j++) {
             Field * f = m->fields[j];
             
             bool ignoreField = f->isAutoFilled();
```
**bool check:** (no suggestions)

davidbrownza replied:
You can actually just remove this line and use `f->isAutoFilled()` in the if-statement.

rowan08 replied:
Done

rowan08 commented on src/model/Model.cpp:
```diff
@@ -198,25 +182,23 @@ void Model::generateSQLParts(string &sqlHead, string &sqlBatch, string &sqlUpdat
     
     bool firstField = true;
```
**bool check:** `isFirstField` ?

davidbrownza replied:
Yup

rowan08 replied:
Done

davidbrownza commented on src/model/Model.cpp:
```diff
@@ -6,93 +6,85 @@ Model::Model(string name) {
     
     string sqlTail = generateSQLTail(sqlBatch, sqlUpdate, batchsize, mode);
     
     //cout << sqlHead + sqlTail << endl;
```
Remove this comment

rowan08 replied:
Done

davidbrownza reviewed (changes requested):
This project is starting to look pretty shlick :)

rowan08 committed:
Changes after CR, round 2:
Made changes to addition boolean variable names
Added boolean naming guidelines to README file.
Removed unwanted comments.
Issue: https://github.com/davidbrownza/CORMAP/issues/34

rowan08 commented on src/fields/Field.cpp:
```diff
@@ -1,42 +1,42 @@
 
-void Field::setNull(bool null) {
-    _isNull = null;
+void Field::setNull(bool isNull) {
```
This was the additional `bool` case I found. I'm assuming the input parameter `isNull` won't interfere with the `isNull()` member function?
